src/inmanta/data/__init__.py

```diff
@@ -6667,59 +6673,88 @@ async def set_last_processed_model_version(
-    host: str,
-    port: int,
-    database: str,
-    username: str,
-    password: str,
-    create_db_schema: bool = True,
-    connection_pool_min_size: int = 10,
-    connection_pool_max_size: int = 10,
-    connection_timeout: float = 60,
-) -> asyncpg.pool.Pool:
-    pool = await asyncpg.create_pool(
-        host=host,
-        port=port,
-        database=database,
-        user=username,
-        password=password,
-        min_size=connection_pool_min_size,
-        max_size=connection_pool_max_size,
-        timeout=connection_timeout,
+def get_async_session(connection_string: typing.Optional[str] = None) -> AsyncSession:
+    if ENGINE is None:
+        raise Exception("Cannot get session because engine wasn't started. Make sure to call start_engine() first.")
+    if ASYNC_SESSION is None:
+        # should not happen
+        raise Exception("Engine is started but session factory wasn't initialized properly.")
+
+    return ASYNC_SESSION()
+
+
+async def stop_engine():
+    global ENGINE
+    await ENGINE.dispose()
+    ENGINE = None
+
+
+def start_engine(
+    url: str,
+    pool_size: int = 10,
+    max_overflow: int = 0,
+    pool_timeout: float = 60.0,
+    echo: bool = False,
+):
+    """
+    engine vs connection vs session overview
+    https://stackoverflow.com/questions/34322471/sqlalchemy-engine-connection-and-session-difference
+    """
+    global ENGINE
+    global ASYNC_SESSION
+    global POOL
+
+    if ENGINE is not None:
+        raise Exception("Engine already running: cannot call start_engine twice.")
+    LOGGER.debug("Creating engine...")
+    ENGINE = create_async_engine(
+        url=url,
+        pool_size=pool_size,
+        max_overflow=max_overflow,
+        pool_timeout=pool_timeout,
+        echo=echo,
     )
-    try:
-        set_connection_pool(pool)
-        if create_db_schema:
-            async with pool.acquire() as con:
-                await schema.DBSchema(CORE_SCHEMA_NAME, PACKAGE_WITH_UPDATE_FILES, con).ensure_db_schema()
-            # expire connections after db schema migration to ensure cache consistency
-            await pool.expire_connections()
-        return pool
-    except Exception as e:
-        await pool.close()
-        await disconnect()
-        raise e
+    POOL = ENGINE.pool
+    ASYNC_SESSION = async_sessionmaker(ENGINE)
+
+    @event.listens_for(ENGINE.sync_engine, "do_connect")
+    def do_connect(dialect, conn_rec, cargs, cparams):
+        print("some-function")
+        print(dialect, conn_rec, cargs, cparams)
+
+    @event.listens_for(ENGINE.sync_engine, "engine_connect")
+    def engine_connect(conn, branch):
+        # print("engine_connect", conn.exec_driver_sql("select 1").scalar())
+        print("engine_connect")
+        print(branch)
+
+    @event.listens_for(ENGINE.sync_engine, "checkout")
+    def my_on_checkout(dbapi_conn, connection_rec, connection_proxy):
+        print("pool checkout")
+        print(dbapi_conn, connection_rec, connection_proxy)
```

Comment:
What are these prints? And can we use the logger instead?

Reply:
I wanted to keep them around to have a readily available example of the event system of sqlalchemy but I removed these since they are not in use.